355: Nonce verification for routes

ocean90 opened the pull request:
See #355.

ocean90 committed:
Routes: Add example for nonce verification

ocean90 commented on gp-includes/routes/settings.php:
```diff
@@ -41,7 +41,7 @@ public function settings_post( $user_id = null ) {
+			if ( ! $this->verify_nonce( 'update-settings_' . $user_id ) ) {
 				gp_notice_set( __( 'An error has occurred. Please try again.', 'glotpress' ), 'error' );
 				$this->redirect( gp_url( '/settings' ) );
 				exit;
```
Alternatively, we could move 45-47 to `verify_nonce()` too, similar to `check_admin_referer()`.

toolstack replied:
Makes sense as we're aborting everything else anyway.  Would it make sense to allow for a `verify_nonce()` call without parameters that could then use the class name and current user to create the nonce id?

toolstack replied:
Maybe make a `$this->verify_nonce()` and `$this->verify_nonce_and_redirect_on_error()` so we can control where the redirect happens.

ocean90 replied:
> call without parameters that could then use the class name and current user to create the nonce id

It's not always the user ID. It can be any ID like for a set, project or translation.

toolstack replied:
That's why it would be an optional format, still allow for the override, but set a default if nothing is passed in.

ocean90 commented on gp-includes/routes/settings.php:
```diff
@@ -41,7 +41,7 @@ public function settings_post( $user_id = null ) {
 
-			if ( empty( $_POST['_wpnonce'] ) || ! wp_verify_nonce( $_POST['_wpnonce'], 'update-settings_' . $user_id ) ) {
+			if ( ! $this->verify_nonce( 'update-settings_' . $user_id ) ) {
 				gp_notice_set( __( 'An error has occurred. Please try again.', 'glotpress' ), 'error' );
```
Should probably be `$this->errors[] = __( 'An error has occurred. Please try again.', 'glotpress' )`.

toolstack replied:
Probably but it does the same effective thing either way.

ocean90 committed:
Routes: Introduce `verify_nonce_and_redirect()`.

Can be used to verify a nonce and redirect on failure.

Also, add `get_http_referer()` and `get_raw_referer()` to retrieve a safe referer source.

toolstack commented on gp-includes/route.php:
```diff
@@ -118,6 +118,56 @@ public function cannot_and_redirect( $action, $object_type = null, $object_id =
+	 */
+	public function verify_nonce_and_redirect( $action, $url = null ) {
+		if ( $this->verify_nonce( $action ) ) {
+			return false;
```
Shouldn't this be `true` to match `veryify_nonce`?

ocean90 replied:
No, because the main action is the redirect.

`cannot_and_redirect()` does the same.

toolstack replied:
Weird, but ok I guess...

ocean90 replied:
Is `nonce_invalid_and_redirect()` a better name?

toolstack replied:
Makes more sense to me, makes it clear the returned false indicates a valid nonce.

ocean90 committed:
Routes: Rename verify_nonce_and_redirect() to invalid_nonce_and_redirect()

ocean90 committed:
Routes: Add nonce to new project form.

ocean90 committed:
Routes: Add nonce to delete project form.

ocean90 committed:
Routes: Add nonce to edit project form.

ocean90 committed:
Routes: Add nonce to new translation set form.

ocean90 committed:
Routes: Add nonce to edit translation set form.

ocean90 committed:
Routes: Add nonce to delete translation form.

ocean90 committed:
Routes: Add nonce to branch project form.

ocean90 committed:
Routes: Add gp_route_nonce_url() for nonces via query args.

ocean90 committed:
Routes: Add nonce to create/delete project permissions form.

ocean90 committed:
Routes: Add nonce to import translations/originals forms.

ocean90 committed:
Routes: Adjust tests to include nonces

ocean90 committed:
Routes: Add nonce to create glossary form.

ocean90 committed:
Routes: Add nonce to edit glossary form.

ocean90 committed:
Docs: Add documentation to `GP` class.

ocean90 committed:
Routes: Add nonce to delete glossary form.

ocean90 committed:
Routes: Add nonce to add glossary entry form.

ocean90 committed:
Routes: Add nonce to import glossary entries form.

ocean90 committed:
Routes: Add nonce to translation status actions.

ocean90 committed:
Routes: Add nonce to discard translation warning action.

ocean90 committed:
Routes: Add nonce to add translation action.

ocean90 committed:
Routes: Add nonce to set original priority action.

ocean90 committed:
Routes: Add nonce to translation bulk actions.

ocean90 committed:
Routes: Add nonce to edit glossary entry action.

ocean90 committed:
Routes: Adjust tests to include nonces.

ocean90 committed:
Routes: Add nonce to delete glossary entry action.

ocean90 committed:
Routes: Add $echo parameter to gp_route_nonce_field().

ocean90 committed:
Routes: Add nonce to personal project settings form.

ocean90 committed:
Routes: Add nonce to mass create translation sets form.
